text/0000-inline-const.md

```diff
@@ -0,0 +1,276 @@
+>    | -? INTEGER_LITERAL
+>    | -? FLOAT_LITERAL
+>    | PathInExpression
+>    | QualifiedPathInExpression
+>    | ConstBlockExpression // new
+> ```
+
+In both the expression and pattern context, an inline `const` behaves exactly
+as if the user had declared a uniquely identified `const` with the block's
+contents as its initializer. For example, in expression context, writing
+`const { ... }` is equivalent to writing:
+
+```rust
+{ const UNIQUE_IDENT: Ty = ...; UNIQUE_IDENT }
+```
+
+where `Ty` is inferred from the expression inside the braces.
+
+An inline `const` is eligible for promotion in an implicit context (just like a
+named `const`), so the following are all guaranteed to work:
+
+```rust
+let x: &'static i32 = &const { 4i32.pow(4) }; // NOT IDIOMATIC
+let x: &'static i32 = const { &4i32.pow(4) }; // IDIOMATIC
+
+// If RFC 2203 is stabilized
+let v = [const { Vec::new() }; 3];  // IDIOMATIC
+let v = const { [ Vec::new(); 3] }; // NOT IDIOMATIC
+```
+
+Whether to lint against the non idiomatic versions is an open question.
+Personally, I would like to lint against `&const {...}` but not `const { [expr;
+3] }`.
+
+Inline `const`s are allowed within `const` and `static` initializers, just as we
+currently allow nested `const` declarations. Whether to lint against inline
+`const` expressions inside a `const` or `static` is also an open question.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This excludes other uses of the `const` keyword in expressions and patterns.
+I'm not aware of any other proposals that would take advantage of this.
+
+This would also be the first use of type inference for const initializers. I'm
+not aware of any technical issues that would arise from this, but perhaps I'm
+overlooking something?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main alternative is the status quo. Maintaining it will likely result in
+promotion being used for more contexts. The lang-team decided to [explore this
+approach](https://github.com/rust-lang/rust/pull/70042#issuecomment-612221597)
+instead.
+
+It would also possible to separate out the parts of this RFC relating to patterns
+so that they can be decided upon seperately. I think they are similar enough
+that they are best considered as a unit, however.
+
+# Prior art
+[prior-art]: #prior-art
+
+I'm not aware of equivalents in other languages.
+
+AFAIK, this was [first proposed] by @scottmcm.
+
+[first proposed]: https://internals.rust-lang.org/t/quick-thought-const-blocks/7803/9
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Naming
+
+I prefer the name inline `const`, since it signals that there is no difference
+between a named `const` and an inline one. @scottmcm prefers "`const` block",
+which is closer to the syntax and parallels the current terminology of `async`
+block and `unsafe` block. It also avoids any accidental conflation with the
+`#[inline]` attribute, which is unrelated.
```

Comment:
The name I used for it so far is "anonymous const", a bit like how closures are sometimes referred to as "anonymous functions".

Reply:
I'll add this to the section.